code/datums/components/shielded.dm

```diff
@@ -0,0 +1,160 @@
+	RegisterSignal(wearer, COMSIG_PARENT_QDELETING, .proc/lost_wearer)
+	if(current_charges)
+		wearer.update_icon()
+
+/// Either we've been dropped or our wearer has been QDEL'd. Either way, they're no longer our problem
+/datum/component/shielded/proc/lost_wearer(datum/source, mob/user)
+	SIGNAL_HANDLER
+
+	if(wearer)
+		UnregisterSignal(wearer, list(COMSIG_ATOM_UPDATE_OVERLAYS, COMSIG_PARENT_QDELETING))
+		wearer.update_icon()
+		wearer = null
+
+/// Used to draw the shield overlay on the wearer
+/datum/component/shielded/proc/on_update_overlays(atom/parent_atom, list/overlays)
+	SIGNAL_HANDLER
+
+	overlays += mutable_appearance(shield_icon_file, (current_charges > 0 ? shield_icon : "broken"), MOB_SHIELD_LAYER)
+
+/**
+ * This proc fires when we're hit, and is responsible for checking if we're charged, then deducting one + returning that we're blocking if so.
+ * It then runs the callback in [/datum/component/shielded/var/on_hit_effects] which handles the messages/sparks (so the visuals)
+ */
+/datum/component/shielded/proc/on_hit_react(datum/source, mob/living/carbon/human/owner, atom/movable/hitby, attack_text, damage, attack_type)
+	SIGNAL_HANDLER
+
+	COOLDOWN_START(src, recently_hit_cd, recharge_start_delay)
+
+	if(current_charges <= 0)
+		return
+	. = COMPONENT_HIT_REACTION_BLOCK
+	current_charges = max(current_charges - 1, 0)
+
+	INVOKE_ASYNC(src, .proc/actually_run_hit_callback, owner, attack_text, current_charges)
+
+	if(!recharge_start_delay) // if recharge_start_delay is 0, we don't recharge
+		if(!current_charges) // obviously if someone ever adds a manual way to replenish charges, change this
+			qdel(src)
+		return
+
+	if(!current_charges)
+		wearer.update_icon()
+	START_PROCESSING(SSdcs, src) // if we DO recharge, start processing so we can do that
+
+/// The wrapper to invoke the on_hit callback, so we don't have to worry about blocking in the signal handler
+/datum/component/shielded/proc/actually_run_hit_callback(mob/living/owner, attack_text, current_charges)
+	on_hit_effects.Invoke(owner, attack_text, current_charges)
+
+/// Default on_hit proc, since cult robes are stupid and have different descriptions/sparks
+/datum/component/shielded/proc/default_run_hit_callback(mob/living/owner, attack_text, current_charges)
+	do_sparks(2, TRUE, owner)
+	owner.visible_message("<span class='danger'>[owner]'s shields deflect [attack_text] in a shower of sparks!<span>")
+	if(current_charges <= 0)
+		owner.visible_message("<span class='warning'>[owner]'s shield overloads!</span>")
+
+/datum/component/shielded/proc/check_recharge_rune(datum/source, obj/item/wizard_armour_charge/recharge_rune, mob/living/user)
+	SIGNAL_HANDLER
+
+	if(!istype(recharge_rune))
+		return
+	. = COMPONENT_NO_AFTERATTACK
+	if(!istype(parent, /obj/item/clothing/suit/space/hardsuit/shielded/wizard))
+		to_chat(user, "<span class='warning'>The rune can only be used on battlemage armour!</span>")
+		return
```

Comment:
This code should be generic, since its part of a generic component. There shouldn't be specific type checking.